tpl/images/images.go

```diff
@@ -117,3 +122,65 @@ func (ns *Namespace) Filter(args ...any) (images.ImageResource, error) {
 
 	return img.Filter(filtersv...)
 }
+
+var (
+	qrErrorCorrectionLevels = map[string]qr.Level{
+		"low":      qr.L,
+		"medium":   qr.M,
+		"quartile": qr.Q,
+		"high":     qr.H,
+	}
+	qrDefaultErrorCorrectLevel = "medium"
+)
+
+// QR encodes the given text to a QR code using the given error correction
+// level, returning an image resource.
+func (ns *Namespace) QR(args ...any) (images.ImageResource, error) {
+	if len(args) == 0 || len(args) > 2 {
+		return nil, errors.New("requires one or two arguments")
+	}
+
+	text, err := cast.ToStringE(args[0])
+	if err != nil {
+		return nil, err
+	}
+	if text == "" {
+		return nil, errors.New("cannot encode an empty string")
+	}
+
+	levels := qrDefaultErrorCorrectLevel
+	if len(args) == 2 {
+		levels, err = cast.ToStringE(args[1])
+		if err != nil {
+			return nil, err
+		}
+		if levels == "" {
+			levels = qrDefaultErrorCorrectLevel
+		}
+	}
+
+	level, ok := qrErrorCorrectionLevels[levels]
+	if !ok {
+		return nil, errors.New("error correction level must be one of low, medium, quartile, or high")
+	}
+
+	code, err := qr.Encode(text, level)
+	if err != nil {
+		return nil, err
+	}
+
+	png := code.PNG()
+	targetPath := fmt.Sprintf("qr_%v.png", hashing.XxHashFromStringHexEncoded(text+levels))
+
+	r, err := ns.createClient.FromString(targetPath, string(png))
+	if err != nil {
+		return nil, err
+	}
+
+	ir, ok := r.(images.ImageResource)
+	if !ok {
+		return nil, errors.New("bug: resource is not an image resource")
```

Comment:
I prefer panics for bugs. In this case it doesn't matter much, as it will look like an error, but at least I wouldn't look for a corresponding test ...

Reply:
I'll make both changes.